internal/pkg/service/stream/storage/repository/slice/slice_state.go

```diff
@@ -0,0 +1,149 @@
+	"github.com/keboola/keboola-as-code/internal/pkg/utils/errors"
+)
+
+func (r *Repository) SwitchToUploading(k model.SliceKey, now time.Time) *op.AtomicOp[model.Slice] {
+	return r.stateTransition(k, now, model.SliceClosing, model.SliceUploading)
+}
+
+func (r *Repository) SwitchToUploaded(k model.SliceKey, now time.Time) *op.AtomicOp[model.Slice] {
+	return r.stateTransition(k, now, model.SliceUploading, model.SliceUploaded)
+}
+
+func (r *Repository) updateSlicesOnFileImport() {
+	r.plugins.Collection().OnFileSave(func(ctx context.Context, now time.Time, original, file *model.File) error {
+		if original != nil && original.State != file.State && file.State == model.FileImported {
+			op.AtomicOpFromCtx(ctx).AddFrom(r.switchSlicesToImported(*file, now))
+		}
+		return nil
+	})
+}
+
+func (r *Repository) validateSlicesOnFileStateTransition() {
+	r.plugins.Collection().OnFileSave(func(ctx context.Context, now time.Time, original, file *model.File) error {
+		if original != nil && original.State != file.State && file.State != model.FileClosing && file.State != model.FileImported {
+			// FileClosing state is handled by the closeSliceOnFileClose method.
+			// FileImported state is handled by the updateSlicesOnFileImport method.
+			op.AtomicOpFromCtx(ctx).AddFrom(r.validateSliceStates(*file))
+		}
+		return nil
+	})
+}
+
+func (r *Repository) stateTransition(k model.SliceKey, now time.Time, from, to model.SliceState) *op.AtomicOp[model.Slice] {
+	var file model.File
+	var old, updated model.Slice
+	return op.Atomic(r.client, &updated).
+		// Read file
+		Read(func(ctx context.Context) op.Op {
+			return r.files.Get(k.FileKey).WithResultTo(&file)
+		}).
+		// Read entity for modification
+		Read(func(ctx context.Context) op.Op {
+			return r.Get(k).WithResultTo(&old)
+		}).
+		// Update the entity
+		Write(func(ctx context.Context) op.Op {
+			return r.switchState(ctx, file.State, old, now, from, to).SetResultTo(&updated)
+		})
+}
+
+func (r *Repository) switchState(ctx context.Context, fileState model.FileState, oldValue model.Slice, now time.Time, from, to model.SliceState) *op.TxnOp[model.Slice] {
+	// Validate from state
+	if oldValue.State != from {
+		return op.ErrorTxn[model.Slice](errors.Errorf(`slice "%s" is in "%s" state, expected "%s"`, oldValue.SliceKey, oldValue.State, from))
+	}
+
+	// Validate file and slice state combination
+	if err := validateFileAndSliceState(fileState, to); err != nil {
+		return op.ErrorTxn[model.Slice](errors.PrefixErrorf(err, `unexpected slice "%s" state:`, oldValue.SliceKey))
+	}
+
+	// Switch slice state
+	newValue, err := oldValue.WithState(now, to)
+	if err != nil {
+		return op.ErrorTxn[model.Slice](err)
+	}
+
+	return r.save(ctx, now, &oldValue, &newValue)
+}
+
+func (r *Repository) switchStateInBatch(ctx context.Context, fState model.FileState, original []model.Slice, now time.Time, from, to model.SliceState) *op.TxnOp[[]model.Slice] {
```

Comment:
`fState` -> `fileState` name in above function is `fileState`.

Reply:
https://github.com/keboola/keboola-as-code/pull/1791/commits/9dd12e86c5d05fa06f252d54d3d2c08e7190f489